File: leetcode/medium/734-sentence-similarity/734-sentence-similarity.kt
class Solution {
    fun areSentencesSimilar(sentence1: Array<String>, sentence2: Array<String>, similarPairs: List<List<String>>): Boolean {
        if (sentence1.size != sentence2.size) {
            return false
        }

        val similar = mutableSetOf<Pair<String, String>>()

        for (pair in similarPairs) {
            similar.add(Pair(pair[0], pair[1]))
            similar.add(Pair(pair[1], pair[0]))
        }

        for ((word1, word2) in sentence1.zip(sentence2)) {
            if (word1 != word2 && !similar.contains(Pair(word1, word2))) {
                return false
            }
        }

        return true
    }
}

File: leetcode/medium/734-sentence-similarity/734-sentence-similarity.py
from typing import List


class Solution:
    def areSentencesSimilar(self, sentence1: List[str], sentence2: List[str], similarPairs: List[List[str]]) -> bool:
        if len(sentence1) != len(sentence2):
            return False

        similar = set()

        for pair in similarPairs:
            similar.add(tuple(pair))
            similar.add((pair[1], pair[0]))

        for word1, word2 in zip(sentence1, sentence2):
            if word1 != word2 and (word1, word2) not in similar:
                return False

        return True
